universal-application-tool-0.0.1/app/services/program/Predicate.java

```diff
@@ -0,0 +1,16 @@
+package services.program;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableSet;
+
+/** String representation of a predicate with variables from the core data model. */
+@AutoValue
+abstract class Predicate {
+  static Predicate create(String function, ImmutableSet<String> variables) {
+    return new AutoValue_Predicate(function, variables);
+  }
+
+  abstract String function();
```

Comment:
Suggestion (blocking): functions are their own concept in the [JsonPath library](https://github.com/json-path/JsonPath#functions). Recommend renaming this to `expression`.

Reply:
done!